```
../librz/core/cio.c
../librz/core/cio.c: In function ‘rz_core_write_base64_at’:
../librz/core/cio.c:822:2: error: ‘buf’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  822 |  free(buf);
      |  ^~~~~~~~~
```

--- a/librz/core/cbin.c
+++ b/librz/core/cbin.c
@@ -5243,3 +5243,22 @@ RZ_API bool rz_core_bin_archs_print(RzBin *bin, RzCmdStateOutput *state) {
 	rz_cmd_state_output_array_end(state);
 	return true;
 }
+
+RZ_API bool rz_core_bin_pdb_load(RZ_NONNULL RzCore *core, RZ_NONNULL const char *filename) {
+	rz_cons_push();
+	rz_core_pdb_info(core, filename, NULL, RZ_MODE_RIZINCMD);
+	const char *buf = rz_cons_get_buffer();
+	if (!buf) {
+		rz_cons_pop();
+		return false;
+	}
+	char *s = strdup(buf);
+	rz_cons_pop();
+	if (!s) {
+		return false;
+	}
+
+	RzCmdStatus status = rz_core_cmd_rzshell(core, s, 0);
+	free(s);
+	return status == RZ_CMD_STATUS_OK;
+}

--- a/librz/core/cio.c
+++ b/librz/core/cio.c
@@ -648,3 +648,220 @@ RZ_API RzCmdStatus rz_core_io_plugins_print(RzIO *io, RzCmdStateOutput *state) {
 	rz_cmd_state_output_array_end(state);
 	return RZ_CMD_STATUS_OK;
 }
+
+/**
+ * \brief Write a given \p value at the specified \p address, using \p sz bytes
+ *
+ * \param core RzCore reference
+ * \param addr Address where to write the value
+ * \param value Value to write
+ * \param sz Number of bytes to write. Can be 1, 2, 4, 8 or the special value 0
+ *           if you want the function to choose based on \p value (4 if \p value
+ *           is <4GB, 8 otherwise)
+ */
+RZ_API bool rz_core_write_value_at(RzCore *core, ut64 addr, ut64 value, int sz) {
+	rz_return_val_if_fail(sz == 0 || sz == 1 || sz == 2 || sz == 4 || sz == 8, false);
+	ut8 buf[sizeof(ut64)];
+	bool be = rz_config_get_i(core->config, "cfg.bigendian");
+
+	core->num->value = 0;
+	if (sz == 0) {
+		sz = value & UT64_32U ? 8 : 4;
+	}
+
+	switch (sz) {
+	case 1:
+		rz_write_ble8(buf, (ut8)(value & UT8_MAX));
+		break;
+	case 2:
+		rz_write_ble16(buf, (ut16)(value & UT16_MAX), be);
+		break;
+	case 4:
+		rz_write_ble32(buf, (ut32)(value & UT32_MAX), be);
+		break;
+	case 8:
+		rz_write_ble64(buf, value, be);
+		break;
+	default:
+		return false;
+	}
+
+	if (!rz_core_write_at(core, addr, buf, sz)) {
+		RZ_LOG_ERROR("Could not write %d bytes at %" PFMT64x "\n", sz, addr);
+		core->num->value = 1;
+		return false;
+	}
+	if (rz_config_get_i(core->config, "cfg.wseek")) {
+		rz_core_seek_delta(core, sz, true);
+	}
+
+	return true;
+}
+
+/**
+ * \brief Write at \p addr the current value + \p value passed as argument
+ *
+ * The values read/written are considered as integers of \p sz bytes.
+ *
+ * \param core RzCore reference
+ * \param addr Address where to overwrite the value
+ * \param value Value to sum to the existing value in \p addr
+ * \param sz Size of the values, in bytes, to consider. Can be 1, 2, 4, 8.
+ */
+RZ_API bool rz_core_write_value_inc_at(RzCore *core, ut64 addr, st64 value, int sz) {
+	rz_return_val_if_fail(sz == 1 || sz == 2 || sz == 4 || sz == 8, false);
+
+	ut8 buf[sizeof(ut64)];
+	bool be = rz_config_get_i(core->config, "cfg.bigendian");
+
+	if (!rz_io_read_at_mapped(core->io, addr, buf, sz)) {
+		return false;
+	}
+
+	switch (sz) {
+	case 1: {
+		ut8 cur = rz_read_ble8(buf);
+		cur += value;
+		rz_write_ble8(buf, cur);
+		break;
+	}
+	case 2: {
+		ut16 cur = rz_read_ble16(buf, be);
+		cur += value;
+		rz_write_ble16(buf, cur, be);
+		break;
+	}
+	case 4: {
+		ut32 cur = rz_read_ble32(buf, be);
+		cur += value;
+		rz_write_ble32(buf, cur, be);
+		break;
+	}
+	case 8: {
+		ut64 cur = rz_read_ble64(buf, be);
+		cur += value;
+		rz_write_ble64(buf, cur, be);
+		break;
+	}
+	default:
+		rz_warn_if_reached();
+		break;
+	}
+
+	if (!rz_core_write_at(core, addr, buf, sz)) {
+		RZ_LOG_ERROR("Could not write %d bytes at %" PFMT64x "\n", sz, addr);
+		return false;
+	}
+	if (rz_config_get_i(core->config, "cfg.wseek")) {
+		rz_core_seek_delta(core, sz, true);
+	}
+
+	return true;
+}
+
+/**
+ * \brief Write a given string \p s at the specified \p addr
+ *
+ * \param core RzCore reference
+ * \param addr Address where to write the string
+ * \param s String to write. The string is unescaped, meaning that if there is `\n` it becomes 0x0a
+ */
+RZ_API bool rz_core_write_string_at(RzCore *core, ut64 addr, const char *s) {
+	rz_return_val_if_fail(core && s, false);
+
+	char *str = strdup(s);
+	if (!str) {
+		return false;
+	}
+
+	int len = rz_str_unescape(str);
+	if (!rz_core_write_at(core, addr, (const ut8 *)str, len)) {
+		RZ_LOG_ERROR("Could not write '%s' at %" PFMT64x "\n", s, addr);
+		return false;
+	}
+	if (rz_config_get_i(core->config, "cfg.wseek")) {
+		rz_core_seek_delta(core, len, true);
+	}
+	return true;
+}
+
+/**
+ * \brief Write a given string \p s at the specified \p addr encoded as base64.
+ *
+ * \param core RzCore reference
+ * \param addr Address where to write the string
+ * \param s String to encode as base64 and then written.
+ */
+RZ_API bool rz_core_write_base64_at(RzCore *core, ut64 addr, const char *s) {
+	rz_return_val_if_fail(core && s, false);
+
+	bool res = false;
+	size_t str_len = strlen(s) + 1;
+	ut8 *bin_buf = malloc(str_len);
+	if (!bin_buf) {
+		return false;
+	}
+
+	const int bin_len = rz_hex_str2bin(s, bin_buf);
+	if (bin_len <= 0) {
+		free(bin_buf);
+		return false;
+	}
+
+	ut8 *buf = calloc(str_len + 1, 4);
+	if (!buf) {
+		free(bin_buf);
+		return false;
+	}
+
+	int len = rz_base64_encode((char *)buf, bin_buf, bin_len);
+	free(bin_buf);
+	if (len == 0) {
+		goto err;
+	}
+
+	if (!rz_core_write_at(core, addr, buf, len)) {
+		RZ_LOG_ERROR("Could not write base64 encoded string '%s' at %" PFMT64x "\n", s, addr);
+		goto err;
+	}
+	if (rz_config_get_i(core->config, "cfg.wseek")) {
+		rz_core_seek_delta(core, len, true);
+	}
+	res = true;
+
+err:
+	free(buf);
+	return res;
+}
+
+/**
+ * \brief Write a given base64 string \p s at the specified \p addr, decoded
+ *
+ * \param core RzCore reference
+ * \param addr Address where to write the string
+ * \param s String to decode from base64 and then written
+ */
+RZ_API bool rz_core_write_base64d_at(RzCore *core, ut64 addr, const char *s) {
+	rz_return_val_if_fail(core && s, false);
+
+	bool res = false;
+	size_t str_len = strlen(s) + 1;
+	ut8 *buf = malloc(str_len);
+	int len = rz_base64_decode(buf, s, -1);
+	if (len < 0) {
+		goto err;
+	}
+
+	if (!rz_core_write_at(core, addr, buf, len)) {
+		RZ_LOG_ERROR("Could not write base64 decoded string '%s' at %" PFMT64x "\n", s, addr);
+		goto err;
+	}
+	if (rz_config_get_i(core->config, "cfg.wseek")) {
+		rz_core_seek_delta(core, len, true);
+	}
+	res = true;
+
+err:
+	free(buf);
+	return res;
+}

--- a/librz/core/cmd_info.c
+++ b/librz/core/cmd_info.c
@@ -525,22 +525,8 @@ RZ_IPI RzCmdStatus rz_cmd_info_pdb_load_handler(RzCore *core, int argc, const ch
 		return RZ_CMD_STATUS_ERROR;
 	}
 
-	rz_cons_push();
-	rz_core_pdb_info(core, filename, NULL, RZ_MODE_RIZINCMD);
+	RzCmdStatus status = bool2status(rz_core_bin_pdb_load(core, filename));
 	free(filename);
-	const char *buf = rz_cons_get_buffer();
-	if (!buf) {
-		rz_cons_pop();
-		return RZ_CMD_STATUS_ERROR;
-	}
-	char *s = strdup(buf);
-	rz_cons_pop();
-	if (!s) {
-		return RZ_CMD_STATUS_ERROR;
-	}
-
-	RzCmdStatus status = rz_core_cmd_rzshell(core, s, 0);
-	free(s);
 	return status;
 }
 

--- a/librz/core/cmd_write.c
+++ b/librz/core/cmd_write.c
@@ -420,6 +420,7 @@ RZ_IPI int rz_wo_handler_old(void *data, const char *input) {
 #define WSEEK(x, y) \
 	if (wseek) \
 	rz_core_seek_delta(x, y, true)
+
 static void rz_cmd_write_value(RzCore *core, const char *input) {
 	int type = 0;
 	ut64 off = 0LL;
@@ -485,128 +486,42 @@ static void rz_cmd_write_value(RzCore *core, const char *input) {
 	}
 }
 
-static RzCmdStatus common_wv_handler(RzCore *core, int argc, const char **argv, int type) {
-	ut64 off = 0LL;
-	ut8 buf[sizeof(ut64)];
-	int wseek = rz_config_get_i(core->config, "cfg.wseek");
-	bool be = rz_config_get_i(core->config, "cfg.bigendian");
-
-	core->num->value = 0;
-	off = rz_num_math(core->num, argv[1]);
-	if (core->file) {
-		rz_io_use_fd(core->io, core->file->fd);
-	}
-
-	ut64 res = rz_io_seek(core->io, core->offset, RZ_IO_SEEK_SET);
-	if (res == UT64_MAX) {
+static RzCmdStatus common_write_value_handler(RzCore *core, const char *valstr, size_t sz) {
+	ut64 value = rz_num_math(core->num, valstr);
+	if (core->num->nc.errors) {
+		RZ_LOG_ERROR("Could not convert argument to number");
 		return RZ_CMD_STATUS_ERROR;
 	}
-	if (type == 0) {
-		type = off & UT64_32U ? 8 : 4;
-	}
 
-	switch (type) {
-	case 1:
-		rz_write_ble8(buf, (ut8)(off & UT8_MAX));
-		break;
-	case 2:
-		rz_write_ble16(buf, (ut16)(off & UT16_MAX), be);
-		break;
-	case 4:
-		rz_write_ble32(buf, (ut32)(off & UT32_MAX), be);
-		break;
-	case 8:
-		rz_write_ble64(buf, off, be);
-		break;
-	}
-
-	if (!rz_io_write(core->io, buf, type)) {
-		cmd_write_fail(core);
-	} else if (wseek) {
-		rz_core_seek_delta(core, type, true);
-	}
-
-	rz_core_block_read(core);
-	return RZ_CMD_STATUS_OK;
+	return rz_core_write_value_at(core, core->offset, value, sz) ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
 RZ_IPI RzCmdStatus rz_write_value_handler(RzCore *core, int argc, const char **argv) {
-	return common_wv_handler(core, argc, argv, 0);
+	return common_write_value_handler(core, argv[1], 0);
 }
 
 RZ_IPI RzCmdStatus rz_write_value1_handler(RzCore *core, int argc, const char **argv) {
-	return common_wv_handler(core, argc, argv, 1);
+	return common_write_value_handler(core, argv[1], 1);
 }
 
 RZ_IPI RzCmdStatus rz_write_value2_handler(RzCore *core, int argc, const char **argv) {
-	return common_wv_handler(core, argc, argv, 2);
+	return common_write_value_handler(core, argv[1], 2);
 }
 
 RZ_IPI RzCmdStatus rz_write_value4_handler(RzCore *core, int argc, const char **argv) {
-	return common_wv_handler(core, argc, argv, 4);
+	return common_write_value_handler(core, argv[1], 4);
 }
 
 RZ_IPI RzCmdStatus rz_write_value8_handler(RzCore *core, int argc, const char **argv) {
-	return common_wv_handler(core, argc, argv, 8);
+	return common_write_value_handler(core, argv[1], 8);
 }
 
 RZ_IPI RzCmdStatus rz_write_base64_encode_handler(RzCore *core, int argc, const char **argv) {
-	int wseek = rz_config_get_i(core->config, "cfg.wseek");
-	const char *str = argv[1];
-	size_t str_len = strlen(str) + 1;
-	ut8 *bin_buf = malloc(str_len);
-	if (!bin_buf) {
-		eprintf("Error: failed to malloc memory");
-		return RZ_CMD_STATUS_ERROR;
-	}
-
-	const int bin_len = rz_hex_str2bin(str, bin_buf);
-	if (bin_len <= 0) {
-		free(bin_buf);
-		return RZ_CMD_STATUS_ERROR;
-	}
-
-	ut8 *buf = calloc(str_len + 1, 4);
-	int len = rz_base64_encode((char *)buf, bin_buf, bin_len);
-	free(bin_buf);
-	if (len == 0) {
-		free(buf);
-		return RZ_CMD_STATUS_ERROR;
-	}
-
-	if (!rz_core_write_at(core, core->offset, buf, len)) {
-		cmd_write_fail(core);
-		free(buf);
-		return RZ_CMD_STATUS_ERROR;
-	}
-	WSEEK(core, len);
-	rz_core_block_read(core);
-
-	free(buf);
-	return RZ_CMD_STATUS_OK;
+	return rz_core_write_base64_at(core, core->offset, argv[1]) ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
 RZ_IPI RzCmdStatus rz_write_base64_decode_handler(RzCore *core, int argc, const char **argv) {
-	int wseek = rz_config_get_i(core->config, "cfg.wseek");
-	const char *str = argv[1];
-	size_t str_len = strlen(str) + 1;
-	ut8 *buf = malloc(str_len);
-	int len = rz_base64_decode(buf, str, -1);
-	if (len < 0) {
-		free(buf);
-		return RZ_CMD_STATUS_ERROR;
-	}
-
-	if (!rz_core_write_at(core, core->offset, buf, len)) {
-		cmd_write_fail(core);
-		free(buf);
-		return RZ_CMD_STATUS_ERROR;
-	}
-	WSEEK(core, len);
-	rz_core_block_read(core);
-
-	free(buf);
-	return RZ_CMD_STATUS_OK;
+	return rz_core_write_base64d_at(core, core->offset, argv[1]) ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
 static bool cmd_wff(RzCore *core, const char *input) {
@@ -1067,7 +982,16 @@ RZ_IPI int rz_w0_handler_old(void *data, const char *input) {
 
 RZ_IPI RzCmdStatus rz_write_zero_handler(RzCore *core, int argc, const char **argv) {
 	ut64 len = rz_num_math(core->num, argv[1]);
-	return rz_cmd_int2status(w0_handler_common(core, len));
+	ut8 *buf = RZ_NEWS0(ut8, len);
+	if (!buf) {
+		RZ_LOG_ERROR("Cannot allocate %" PFMT64d " bytes", len);
+		return RZ_CMD_STATUS_ERROR;
+	}
+
+	bool res = rz_core_write_at(core, core->offset, buf, len);
+	free(buf);
+
+	return res ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
 static int rz_w_incdec_handler_old(void *data, const char *input, int inc) {
@@ -1095,8 +1019,7 @@ static RzCmdStatus w_incdec_handler(RzCore *core, int argc, const char **argv, i
 	if (command[strlen(command) - 1] == '-') {
 		num *= -1;
 	}
-	cmd_write_inc(core, inc_size, num);
-	return RZ_CMD_STATUS_OK;
+	return rz_core_write_value_inc_at(core, core->offset, num, inc_size) ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
 RZ_IPI RzCmdStatus rz_write_1_inc_handler(RzCore *core, int argc, const char **argv) {
@@ -1569,10 +1492,7 @@ RZ_IPI int rz_w_handler_old(void *data, const char *input) {
 }
 
 RZ_IPI RzCmdStatus rz_write_handler(RzCore *core, int argc, const char **argv) {
-	char *s = rz_str_array_join(argv + 1, argc - 1, " ");
-	w_handler_common(core, s);
-	free(s);
-	return RZ_CMD_STATUS_OK;
+	return rz_core_write_string_at(core, core->offset, argv[1]) ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
 RZ_IPI int rz_wz_handler_old(void *data, const char *input) {

--- a/librz/core/core_private.h
+++ b/librz/core/core_private.h
@@ -78,7 +78,6 @@ RZ_IPI void rz_core_types_struct_print_format_all(RzCore *core);
 RZ_IPI void rz_core_types_union_print_format_all(RzCore *core);
 RZ_IPI void rz_core_types_link_print(RzCore *core, RzType *type, ut64 addr, RzOutputMode mode, PJ *pj);
 RZ_IPI void rz_core_types_link_print_all(RzCore *core, RzOutputMode mode);
-RZ_IPI void rz_core_types_link(RzCore *core, const char *typestr, ut64 addr);
 RZ_IPI void rz_core_types_link_show(RzCore *core, ut64 addr);
 RZ_IPI void rz_core_types_print_all(RzCore *core, RzOutputMode mode);
 RZ_IPI void rz_types_define(RzCore *core, const char *type);

--- a/librz/core/ctypes.c
+++ b/librz/core/ctypes.c
@@ -966,7 +966,16 @@ RZ_IPI void rz_core_types_link_print_all(RzCore *core, RzOutputMode mode) {
 	}
 }
 
-RZ_IPI void rz_core_types_link(RzCore *core, const char *typestr, ut64 addr) {
+/**
+ * \brief Link an address \p addr to the type referenced by \p typestr
+ *
+ * NOTE: This is likely going to be deprecated with the use of global variables.
+ *
+ * \param core RzCore reference
+ * \param typestr Name of the type that should be defined at \p addr
+ * \param addr Address where the type should be used
+ */
+RZ_API void rz_core_types_link(RzCore *core, const char *typestr, ut64 addr) {
 	char *error_msg = NULL;
 	RzType *type = rz_type_parse_string_single(core->analysis->typedb->parser, typestr, &error_msg);
 	if (!type || error_msg) {

--- a/librz/include/rz_core.h
+++ b/librz/include/rz_core.h
@@ -568,6 +568,11 @@ RZ_API int rz_core_file_binlist(RzCore *core);
 RZ_API bool rz_core_file_bin_raise(RzCore *core, ut32 num);
 RZ_API bool rz_core_extend_at(RzCore *core, ut64 addr, int size);
 RZ_API bool rz_core_write_at(RzCore *core, ut64 addr, const ut8 *buf, int size);
+RZ_API bool rz_core_write_value_at(RzCore *core, ut64 addr, ut64 value, int sz);
+RZ_API bool rz_core_write_value_inc_at(RzCore *core, ut64 addr, st64 value, int sz);
+RZ_API bool rz_core_write_string_at(RzCore *core, ut64 addr, RZ_NONNULL const char *s);
+RZ_API bool rz_core_write_base64d_at(RzCore *core, ut64 addr, RZ_NONNULL const char *s);
+RZ_API bool rz_core_write_base64_at(RzCore *core, ut64 addr, RZ_NONNULL const char *s);
 RZ_API int rz_core_write_op(RzCore *core, const char *arg, char op);
 RZ_API ut8 *rz_core_transform_op(RzCore *core, const char *arg, char op);
 RZ_API ut32 rz_core_file_cur_fd(RzCore *core);
@@ -854,6 +859,7 @@ RZ_API RZ_OWN char *rz_core_bin_field_build_flag_name(RZ_NONNULL RzBinClass *cls
 RZ_API char *rz_core_bin_method_flags_str(ut64 flags, int mode);
 RZ_API bool rz_core_pdb_info(RzCore *core, const char *file, PJ *pj, int mode);
 RZ_API RZ_OWN char *rz_core_bin_pdb_get_filename(RZ_NONNULL RzCore *core);
+RZ_API bool rz_core_bin_pdb_load(RZ_NONNULL RzCore *core, RZ_NONNULL const char *filename);
 RZ_API RzCmdStatus rz_core_bin_plugins_print(RzBin *bin, RzCmdStateOutput *state);
 
 RZ_API bool rz_core_bin_archs_print(RZ_NONNULL RzBin *bin, RZ_NONNULL RzCmdStateOutput *state);

--- a/librz/include/rz_core.h
+++ b/librz/include/rz_core.h
@@ -531,6 +531,7 @@ RZ_API void rz_core_link_stroff(RzCore *core, RzAnalysisFunction *fcn);
 RZ_API bool cmd_analysis_objc(RzCore *core, bool auto_analysis);
 RZ_API void rz_core_analysis_cc_init(RzCore *core);
 RZ_API void rz_core_analysis_paths(RzCore *core, ut64 from, ut64 to, bool followCalls, int followDepth, bool is_json);
+RZ_API void rz_core_types_link(RzCore *core, const char *typestr, ut64 addr);
 
 RZ_API RzListInfo *rz_listinfo_new(const char *name, RzInterval pitv, RzInterval vitv, int perm, const char *extra);
 RZ_API void rz_listinfo_free(RzListInfo *info);

